Sources/SwiftParser/Parser.swift

```diff
@@ -629,3 +658,69 @@ extension Parser {
     )
   }
 }
+
+// MARK: Incremental Parsing
+extension Parser {
+  mutating func loadCurrentSyntaxNodeFromCache(for kind: SyntaxKind) -> Syntax? {
+    guard parseLookup != nil else {
+      return nil
+    }
+
+    let currentOffset = self.lexemes.getOffsetToStart(self.currentToken)
+    if let node = parseLookup!.lookUp(currentOffset, kind: kind) {
+      self.lexemes.advance(by: node.byteSize, currentToken: &self.currentToken)
+      return node
+    }
+
+    return nil
+  }
+
+  mutating func registerNodeForIncrementalParse(node: RawSyntax, startToken: Lexer.Lexeme) {
+    lookaheadRanges.registerNodeForIncrementalParse(
+      node: node,
+      lookaheadLength: lexemes.lookaheadTracker.pointee.furthestOffset - self.lexemes.getOffsetToStart(startToken)
+    )
+  }
+}
+
+/// Record the furthest offset to `sourceBufferStart` that is reached by  `Parser.peek()` or `Parser.Lookahead` in ``Lexer/LexemeSequence``
+public struct LookaheadTracker {
+  private(set) var furthestOffset: Int = 0
+
+  public init() {}
+
+  public mutating func recordFurthestOffset(_ furthestOffset: Int) {
+    /// We could lookahead multi-times to find different valid part of a node, so we should take the maximum of the lookahead offset as the possible affect range of a node.
+    self.furthestOffset = max(furthestOffset, self.furthestOffset)
+  }
+}
+
+/// Owns a ``LookaheadTracker``.
+///
+/// Once the `LookeaheadTrackerOwner` is deinitialized, the ``LookaheadTracker`` is also destroyed.
+class LookaheadTrackerOwner {
+  var lookaheadTracker: UnsafeMutablePointer<LookaheadTracker>
+
+  init() {
+    self.lookaheadTracker = .allocate(capacity: 1)
+    self.lookaheadTracker.initialize(to: LookaheadTracker())
+  }
+
+  deinit {
+    self.lookaheadTracker.deallocate()
+  }
+}
+
+/// Record the lookahead ranges for syntax nodes.
+public struct LookaheadRanges {
+  /// For each node that is recorded for re-use, the number of UTF-8 bytes that the parser looked ahead to parse the node, measured from the start of the node’s leading trivia.
+  /// This information can be used to determine whether a node can be reused in incremental parse. A node can only be re-used if no byte in its looked range has changed.
```

Comment:
I forgot this in my suggestion as well but we should should have a newline between the summary and discussion for this property.

```suggestion
  /// For each node that is recorded for re-use, the number of UTF-8 bytes that the parser looked ahead to parse the node, measured from the start of the node’s leading trivia.
  ///
  /// This information can be used to determine whether a node can be reused in incremental parse. A node can only be re-used if no byte in its looked range has changed.
```